apps/aecore/src/aec_peer_connection.erl

```diff
@@ -0,0 +1,825 @@
+        Ref when NonRegAccept ->
+            lager:info("Connecting peer never sent ping, stopping"),
+            %% Consider blocking this peer
+            close_connection(S),
+            {stop, normal, S};
+        _ ->
+            lager:debug("Got stale first_ping_timeout", []),
+            {noreply, S}
+    end;
+handle_info({connected, Pid, Err = {error, _}}, S = #{ status := {connecting, Pid}}) ->
+    lager:debug("Failed to connected to ~p: ~p", [{maps:get(host, S), maps:get(port, S)}, Err]),
+    connect_fail(S);
+handle_info({connected, Pid, {ok, TcpSock}}, S0 = #{ status := {connecting, Pid} }) ->
+    S = #{ version := Version, genesis := Genesis,
+       seckey := SecKey, pubkey := PubKey, r_pubkey := RemotePub } = ensure_genesis(S0),
+    NoiseOpts = [ {noise, <<"Noise_XK_25519_ChaChaPoly_BLAKE2b">>}
+                , {s, enoise_keypair:new(dh25519, SecKey, PubKey)}
+                , {prologue, <<Version/binary, Genesis/binary>>}
+                , {rs, enoise_keypair:new(dh25519, RemotePub)}
+                , {timeout, ?NOISE_HS_TIMEOUT} ],
+    case enoise:connect(TcpSock, NoiseOpts) of
+        {ok, ESock, _} ->
+            lager:debug("Peer connected to ~p", [{maps:get(host, S), maps:get(port, S)}]),
+            case aec_peers:connect_peer(peer_id(S), self()) of
+                ok ->
+                    {noreply, S#{ status => {connected, ESock} }};
+                {error, _} ->
+                    lager:debug("Dropping unnecessary connection to ", [maps:get(host, S)]),
+                    enoise:close(ESock),
+                    {stop, normal, S}
+            end;
+        {error, Reason} ->
+            %% For now lets report connect_fail
+            %% Maybe instead report permanent fail + block?!
+            lager:info("Connection handshake failed - ~p was from ~p",
+                       [Reason, maps:get(host, S)]),
+            gen_tcp:close(TcpSock),
+            connect_fail(S)
+    end;
+
+handle_info({noise, _, <<?MSG_FRAGMENT:16, N:16, M:16, Fragment/binary>>}, S) ->
+    handle_fragment(S, N, M, Fragment);
+handle_info({noise, _, <<Type:16, Payload/binary>>}, S) ->
+    {ok, Msg} = msgpack:unpack(Payload),
+    S1 = handle_msg(S, Type, Msg),
+    {noreply, S1};
+handle_info({tcp_closed, _}, S) ->
+    lager:debug("Peer connection got tcp_closed", []),
+    case is_non_registered_accept(S) of
+        true ->
+            {stop, normal, S};
+        false ->
+            connect_fail(S)
+    end;
+handle_info(_Msg, S) ->
+    {noreply, S}.
+
+terminate(_Reason, _State) ->
+    ok.
+
+code_change(_OldVsn, State, _Extra) ->
+    {ok, State}.
+
+%% -- Local functions --------------------------------------------------------
+get_request(S, Kind) ->
+    maps:get(Kind, maps:get(requests, S, #{}), none).
+
+set_request(S, Kind, From) ->
+    Rs = maps:get(requests, S, #{}),
+    S#{ requests => Rs#{ Kind => {Kind, From} } }.
+
+drop_request(S, Kind) ->
```

Comment:
`drop_request` sounds as if it wasn't handled. I think this could be better named for readability.